[READY] Rewrite completion system

micbou opened the pull request:
There is a number of issues with the current completion system:
 - UI is blocked until the completions are returned by the server, the request timed out, or a key is pressed by the user. This leads to a rough experience when completions take too much time: cursor disappearing and timeout errors (see https://github.com/Valloric/YouCompleteMe/issues/2192 and https://github.com/Valloric/YouCompleteMe/issues/2574). Users even [increase the timeout by manually editing the `completion_request.py` file](https://github.com/Valloric/YouCompleteMe/blob/master/python/ycm/client/completion_request.py#L30) to avoid these errors, which exacerbate the issue and, in some cases, make the plugin unusable (see https://github.com/Valloric/YouCompleteMe/issues/2574).
 - no fuzzy matching from omnifunc when forcing semantic completion. See https://github.com/Valloric/YouCompleteMe/issues/961;
 - no fuzzy matching when deleting characters. Vim filtering is used instead:

![completion-bug-deleting-characters](https://cloud.githubusercontent.com/assets/10026824/26276156/f298c6de-3d71-11e7-92da-d22186239c27.gif)
Neovim and MacVim are not affected.
 - completion menu disappears after deleting a character and inserting one:

![completion-bug-reinserting-character](https://cloud.githubusercontent.com/assets/10026824/26276192/b3ed0f7a-3d72-11e7-8c64-523a0a59cbdc.gif)
Neovim and MacVim are not affected.
 - ignore the start column returned by the server. See PR https://github.com/Valloric/YouCompleteMe/pull/2489.
 - subject to flickers. This one depends a lot on the version of Vim. Completion is almost flicker-free in Neovim and MacVim. Not too bad in console Vim (except in fast terminal like [alacritty](https://github.com/jwilm/alacritty)). Awful in gVim GTK2 (a bit better on GTK3) and gVim on Windows.

This PR is an attempt at fixing all of these issues while reducing flickers as best as possible (due to how completion works in Vim, a flicker-free experience is impossible to achieve). Here's how:
 - poll for completions using a timer and call `completefunc` once the completions are ready. Use the start column returned by the server in `completefunc`;
 - immediately display the last completions on the `TextChangedI` event to prevent the popup menu disappearing while waiting for the completions. This reduces flickering;
 - use the `InsertCharPre` event to close the completion menu just before inserting a character. This way the `TextChangedI` event is triggered when the character is inserted (this event is not fired when the completion menu is visible). This replaces the `refresh` option set to `always` in `completefunc` and the `s:cursor_moved` hack;
 - remap the backspace key to close the completion menu when deleting a character and thus triggering the `TextChangedI` event;
 - send a request with `force_semantic` set to `True` when forcing semantic completion instead of calling the omnifunc. This fixes the issue where there is no fuzzy matching for custom omnifuncs.

Here's a demo where I added a spin animation on the command line while loading the completions to show that it doesn't block the Vim UI:

![async-completion-for-real](https://cloud.githubusercontent.com/assets/10026824/26277295/0f16a718-3d86-11e7-90f3-8a56bbf53f9f.gif)

Fixes #961.
Fixes #1282.
Fixes #1881. 
Fixes #2192.
Fixes #2500.
Fixes #2574.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2657)
<!-- Reviewable:end -->

Valloric commented:
Thanks for working on this! Reworking the completion system is a daunting task indeed.

I've given this a spin on gvim and console vim on Ubuntu 16.04 and can confirm that the bugs are fixed. Flicker remains unchanged though to my eyes; both `master` and this PR have quite a bit of it in both gvim and console vim. So at least it doesn't make it worse. :)

Given that it fixes all the other issues, I'm more than happy to see this merged (from a functionality perspective). Haven't look at the code yet though... :)

Valloric commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 240 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkahLtsFej08Ze8tIc1:-KkahLtsFej08Ze8tIc2:b-b9idwo) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L240)):*
> ```vim
> 
>   " Remap the backspace key to refresh completions when deleting a character.
>   " TODO: should we remap <C-h> too? Should we add an option to specify which
> ```

I'd say: yes to the first, no to the second. The first answer might be debatable, but the second one I'm more confident in.

---

*[autoload/youcompleteme.vim, line 676 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkahtC2-aZyz8GNENw0:-KkahtC3kG2D0e3mei15:b-gcwesg) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L676)):*
> ```vim
> 
>   call s:PollCompletion()
>   return ''
> ```

Why does this return an empty string?

---

*[autoload/youcompleteme.vim, line 690 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkahy7D1UDGHEx4egs4:-Kkahy7D1UDGHEx4egs5:b2zl17l) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L690)):*
> ```vim
> function! s:OnBackspaceKey()
>   " Using s:StopCompletion here doesn't work.
>   " TODO: explain why.
> ```

Yes, please. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc50bnWiBhe43Da6Aw:-Kkc50bnWiBhe43Da6Ax:bco1dw4) (no related file):*
For posterity, from gitter seems to trigger semantic completion on a whitespace character in the following test:

i can also confirm that it fixes Valloric/ycmd#671 when used with Valloric/ycmd#681

it's triggering completion after space

![YCM-new-completion.gif](https://files.reviewableusercontent.io/images/reviewable/10584846/p5ftW4qa8IG3IZG3pdJdLw3w/YCM-new-completion.gif)

i hit ' then <c-space> then <space>

here's the old behaviour

![YCM-old-completion.gif](https://files.reviewableusercontent.io/images/reviewable/10584846/6vblfgNGhkR0YeEvGaSzhOmQ/YCM-old-completion.gif)

in practice, it just looks like a flicker because you would just type ) and get on with your life, but thought i'd mention it

here is the `.tarn-project`

```
{
  "plugins": {
    "node": {}
  }
}
```

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc5cV3GhPHMyKI4wom:-Kkc5cV3GhPHMyKI4won:b-wzv6pw) (no related file):*
For posterity: from gitter.

When a completion request times out, you get tracebacks in Vim, which we want to avoid:

```

ConnectionError: ('Connection aborted.', BadStatusLine("''",))
Error detected while processing function <SNR>125_PollCompletion[8]..<SNR>125_Pyeval:
line    4:
E859: Failed to convert returned python object to vim value
Error detected while processing function <SNR>125_PollCompletion:
line    9:
E121: Undefined variable: completion_start_column
E15: Invalid expression: {   'start_column': response.completion_start_column,   'candidates': response.completions }
ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=50562): Max retries exceeded with url: /completions (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x10a459c50>: Failed to establish a new connection: [Errno 61] ...
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
I've given this a try. Just like @valloric I can confirm all the mentioned bugs are fixed, except the one that has to do with strts column. Unlike @valloric I think I can see an improvement in flickering.

I've tested on a "small" C++ file that looks like this:
```cpp
#include <boost/bimap.hpp>
#include <boost/python.hpp>

struct a_struct {
	int a;
	int b;
	int c;
	float x;
	float y;
	float z;
	char very_long_char;
};

int main(void) {
	a_struct very_long_struct_name;
	very_long_struct_name.
	return 0;
}
```

The included headers are there just to make compiling take a while.

Here's my `.ycm_extra_conf.py`:
```py
def FlagsForFile(filename):
    return { 'flags' : ['-xc++',
                        '-std=c++1z',
                        '-isystem',
                        '/usr/include/c++/6.3.1',
                        '-isystem',
                        '/usr/include/c++/6.3.1/x86_64-pc-linux-gnu',
                        '-isystem',
                        '/usr/include/c++/6.3.1/backward',
                        '-isystem',
                        '/usr/lib/clang/3.9.1/include',
                        '-isystem',
                        '/usr/include',
                        '-Wall',
                        '-Wextra',
                        '-pedantic',
                        '-Wno-main-return-type',
                        ]}
```

Uhm, please don't ask me about the `-isystem` flags.

> I added a spin animation on the command line while loading the completions to show that it doesn't block the Vim UI

I believe it would be a good idea to leave it in. As a YCM user for a few years, my instinct tells me UI is blocked. So a discreete reminder would be nice.

Just a spinning line may not be enough in my opinion. How about something like `Waiting for completion results` followed by thte same spinning line? Though my suggestion may be a bit too long.

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 240 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkahLtsFej08Ze8tIc1:-Kkc0ezmbvKQ9egKS2lQ:bs34rnq) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L240)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

I'd say: yes to the first, no to the second. The first answer might be debatable, but the second one I'm more confident in.
</blockquote></details>

My first thought was, since `<CR>` and `<C-h>` have the same key codes, we should treat them the same way. But then, upon further thinking, I'm with @valloric. I can't believe I forgot about my own use case. Those two need not have the same key code and they don't on my system, running Suckless terminal and plain old vim in terminal. I even have remapped `<C-h>` to move to the split or tmux pane to the right.

That said, Neovim does behave differently than Vim and does treat those two the same, so I'm not sure what's the smart thing to do here.

For further reading on `<C-h>` differences between Vim and Neovim [click here](https://github.com/christoomey/vim-tmux-navigator/blob/master/README.md#it-doesnt-work-in-neovim-specifically-c-h).

---

*[autoload/youcompleteme.vim, line 616 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc3AEl82mYclwB4x3S:-Kkc3AEl82mYclwB4x3T:b-ggxkkt) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L616)):*
> ```vim
>   " semantic trigger. If we don't, the completion menu stays open while waiting
>   " for the server response.
>   let s:completion = s:default_completion
> ```

I think we should keep this. I believe that keeping the previous menu until the server responds with a new set of semantic completions would be more confusing to the end users as opposed to how much one flicker would annoy them. 

---

*[autoload/youcompleteme.vim, line 690 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkahy7D1UDGHEx4egs4:-Kkc54TWWjG54mbtKecK:b-rz5hso) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L690)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Yes, please. :)
</blockquote></details>

Heh, because here we don't want to throw away the completed text, rather just delete a character.

`s:StopCompletion()` closes the pop-up menu and throw away the completed text with `<C-e>`, while here we close pop-up menu without throwing away the completed text using `<C-y>`.

P.S. Vimscrip is a beautiful language!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-Kkc9fapqlkgNF0FMAue:bscimh9)*
<!-- Sent from Reviewable.io -->

micbou commented:
Thanks for testing and reviewing. I've pushed some changes.

> I believe it would be a good idea to leave it in. As a YCM user for a few years, my instinct tells me UI is blocked. So a discreete reminder would be nice.

> Just a spinning line may not be enough in my opinion. How about something like Waiting for completion results followed by thte same spinning line? Though my suggestion may be a bit too long.

Some users will probably want to disable it or customize it so I think it would be better to define an API for that. In fact, it's already possible to implement your own spin animation by accessing the `ycm_state` object. That's the code I wrote in a `vimrc` for the demo:
```viml
let s:spin_symbols = [ "-", "\\", "|", "/" ]
let s:spin_index = 0
let s:spinner = s:spin_symbols[ s:spin_index ]

function! StartSpinning()
    let s:loading_index = 0
    let s:timer_id = timer_start( 100, "Spin" )
endfunction

function! Spin( ... )
    if !py3eval( "ycm_state.CompletionRequestReady()" )
        echo s:spinner
        let s:spin_index = s:spin_index + 1
        if s:spin_index >= len( s:spin_symbols )
            let s:spin_index = 0
        endif
        let s:spinner = get( s:spin_symbols, s:spin_index )
        let s:timer_id = timer_start( 100, "Spin" )
    else
        echo ''
    endif
endfunction

autocmd TextChangedI * call StartSpinning()
```

---

Reviewed 8 of 9 files at r1, 1 of 2 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc50bnWiBhe43Da6Aw:-KkexG-rhogWwId44qaq:bk637rx) (no related file):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

For posterity, from gitter seems to trigger semantic completion on a whitespace character in the following test:

i can also confirm that it fixes Valloric/ycmd#671 when used with Valloric/ycmd#681

it's triggering completion after space

![YCM-new-completion.gif](https://files.reviewableusercontent.io/images/reviewable/10584846/p5ftW4qa8IG3IZG3pdJdLw3w/YCM-new-completion.gif)

i hit ' then <c-space> then <space>

here's the old behaviour

![YCM-old-completion.gif](https://files.reviewableusercontent.io/images/reviewable/10584846/6vblfgNGhkR0YeEvGaSzhOmQ/YCM-old-completion.gif)

in practice, it just looks like a flicker because you would just type ) and get on with your life, but thought i'd mention it

here is the `.tarn-project`

```
{
  "plugins": {
    "node": {}
  }
}
```


</blockquote></details>

It seems to be fixed by leaving semantic mode and resetting completions before the first `s:Complete()` call. Could you try again with the new changes?

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc5cV3GhPHMyKI4wom:-Kkey6mxYtiQjPRh-llH:b-j8njgf) (no related file):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

For posterity: from gitter.

When a completion request times out, you get tracebacks in Vim, which we want to avoid:

```

ConnectionError: ('Connection aborted.', BadStatusLine("''",))
Error detected while processing function <SNR>125_PollCompletion[8]..<SNR>125_Pyeval:
line    4:
E859: Failed to convert returned python object to vim value
Error detected while processing function <SNR>125_PollCompletion:
line    9:
E121: Undefined variable: completion_start_column
E15: Invalid expression: {   'start_column': response.completion_start_column,   'candidates': response.completions }
ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=50562): Max retries exceeded with url: /completions (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x10a459c50>: Failed to establish a new connection: [Errno 61] ...
```
</blockquote></details>

Strange, the `completion_start_column` key should always exist. Could you check the value of the `response` variable returned by [`GetCompletionResponse`](https://github.com/micbou/YouCompleteMe/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L707) when the error occurs?

---

*[autoload/youcompleteme.vim, line 240 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkahLtsFej08Ze8tIc1:-Kkez9wdVGJ6wnUvb_Kk:b-74uis9) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L240)):*
<details><summary><i>Previously, bstaletic wrote…</i></summary><blockquote>

My first thought was, since `<CR>` and `<C-h>` have the same key codes, we should treat them the same way. But then, upon further thinking, I'm with @valloric. I can't believe I forgot about my own use case. Those two need not have the same key code and they don't on my system, running Suckless terminal and plain old vim in terminal. I even have remapped `<C-h>` to move to the split or tmux pane to the right.

That said, Neovim does behave differently than Vim and does treat those two the same, so I'm not sure what's the smart thing to do here.

For further reading on `<C-h>` differences between Vim and Neovim [click here](https://github.com/christoomey/vim-tmux-navigator/blob/master/README.md#it-doesnt-work-in-neovim-specifically-c-h).
</blockquote></details>

I added `<C-h>` to the list of keys. I also added the `<silent>` attribute so that if the user has already mapped the key to something else, we don't override it.

---

*[autoload/youcompleteme.vim, line 616 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc3AEl82mYclwB4x3S:-Kkf-o4BjRkNGpSlOEw6:b-2ya6r5) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L616)):*
<details><summary><i>Previously, bstaletic wrote…</i></summary><blockquote>

I think we should keep this. I believe that keeping the previous menu until the server responds with a new set of semantic completions would be more confusing to the end users as opposed to how much one flicker would annoy them. 
</blockquote></details>

I agree but more importantly, resetting completions here is needed if we don't want to display the previous ones after entering a new line. I removed the comment.

---

*[autoload/youcompleteme.vim, line 676 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkahtC2-aZyz8GNENw0:-Kkf2yb7M9y48A72uZgg:b-qsimk0) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L676)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Why does this return an empty string?
</blockquote></details>

This function is called in a mapping through the expression register `<C-R>=` (`:h c_CTRL-R_=`) which inserts its return value. We don't insert anything if we return an empty string. Added a comment.

---

*[autoload/youcompleteme.vim, line 690 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkahy7D1UDGHEx4egs4:-Kkf5hMOeUpJZI7QrfX0:b-dbbvha) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L690)):*
<details><summary><i>Previously, bstaletic wrote…</i></summary><blockquote>

Heh, because here we don't want to throw away the completed text, rather just delete a character.

`s:StopCompletion()` closes the pop-up menu and throw away the completed text with `<C-e>`, while here we close pop-up menu without throwing away the completed text using `<C-y>`.

P.S. Vimscrip is a beautiful language!
</blockquote></details>

That's exactly why.

---

*[python/ycm/youcompleteme.py, line 278 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkfiCTxbuGi2GVm02Z8:-KkfiCTxbuGi2GVm02Z9:b-5e17ts) ([raw file](https://github.com/valloric/youcompleteme/blob/40743333f2cd922b2a9f855547cc6715e2b051cb/python/ycm/youcompleteme.py#L278)):*
> ```Python
> 
> 
>   def SendCompletionRequest( self, force_semantic = False ):
> ```

I renamed the function because we now actually send the request.

---

*[python/ycm/youcompleteme.py, line 299 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkfi87y8MmZ0CfQghqD:-Kkfi87y8MmZ0CfQghqE:b-ejek68) ([raw file](https://github.com/valloric/youcompleteme/blob/40743333f2cd922b2a9f855547cc6715e2b051cb/python/ycm/youcompleteme.py#L299)):*
> ```Python
>   def CompletionRequestReady( self ):
>     return bool( self._latest_completion_request and
>                  self._latest_completion_request.Done() )
> ```

I renamed the function to be consistent with `FileParseRequestReady`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-Kkex3shYa4vFjb1tHKh:b-6tjkz3)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 278 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkfiCTxbuGi2GVm02Z8:-KkgCzApvlIeJIZhRAEx:b-kumt64) ([raw file](https://github.com/valloric/youcompleteme/blob/40743333f2cd922b2a9f855547cc6715e2b051cb/python/ycm/youcompleteme.py#L278)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

I renamed the function because we now actually send the request.
</blockquote></details>

Ha, I was mulling over should I point that out or not. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc5cV3GhPHMyKI4wom:-KkgZKgpAMfhQ1Aj14gK:b8eqb37) (no related file):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Strange, the `completion_start_column` key should always exist. Could you check the value of the `response` variable returned by [`GetCompletionResponse`](https://github.com/micbou/YouCompleteMe/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L707) when the error occurs?
</blockquote></details>

If i add a `echom` to print the value, I get

```viml
  let response = s:Pyeval( 'ycm_state.GetCompletionResponse()' )
  echom 'response.completions = ' . response.completions
...
```


```
E859: Failed to convert returned python object to vim value
```

I think the issue is something like: `GetCompletionResponse` raises an exception, and returns "nothing"? Vim can't convert that into anything like a dict, so errors and doesn't set it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc5cV3GhPHMyKI4wom:-Kkge18BkMMRw1LF9U7l:bskcmnj) (no related file):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

If i add a `echom` to print the value, I get

```viml
  let response = s:Pyeval( 'ycm_state.GetCompletionResponse()' )
  echom 'response.completions = ' . response.completions
...
```


```
E859: Failed to convert returned python object to vim value
```

I think the issue is something like: `GetCompletionResponse` raises an exception, and returns "nothing"? Vim can't convert that into anything like a dict, so errors and doesn't set it?

</blockquote></details>

OK i got the value printed by doing this:

```viml
  exec s:python_until_eof

import logging
response_p = ycm_state.GetCompletionResponse()
print( 'response: {0}'.format( response_p ) )
EOF

  let response = s:Pyeval( 'response_p' )
```

```
ConnectionError: ('Connection aborted.', BadStatusLine("''",))
response: {u'errors': [{u'exception': {u'request': {u'body': u'{"column": 8, "file_name": "/Users/ben/Development/Swift/Test/SwiftTest/
SwiftTest/ViewController.swift", "line": 16, "flags": ["/Users/ben/Development/Swift/Test/SwiftTest/SwiftTest/AppDelegate.swift", "-tar
get", "x86_64-apple-macosx10.12", "-enable-objc-interop", "-sdk", "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform
/Developer/SDKs/MacOSX10.12.sdk", "-I", "/Users/ben/Development/Swift/Test/SwiftTest/build/Release", "-F", "/Users/ben/Development/Swif
t/Test/SwiftTest/build/Release", "-g", "-serialize-debugging-options", "-O", "-module-name", "SwiftTest", "-emit-objc-header-path", "/U
sers/ben/Development/Swift/Test/SwiftTest/build/SwiftTest.build/Release/SwiftTest.build/Objects-normal/x86_64/SwiftTest-Swift.h", "-num
-threads", "8"], "contents": "//\\n//  ViewController.swift\\n//  SwiftTest\\n//\\n//  Created by Ben Jackson on 20/05/2017.\\n//  Copy
right \\u00a9 2017 Ben Jackson. All rights reserved.\\n//\\n\\nimport Cocoa\\n\\nclass ViewController: NSViewController {\\n\\n    over
ride func viewDidLoad() {\\n        super.viewDidLoad()\\n\\n        b\\n    }\\n\\n    @IBOutlet weak var button_: NSButton!\\n}\\n\\n
"}', u'url': u'http://127.0.0.1:56334/completions', u'TYPE': u'PreparedRequest', u'headers': {u'_store': {u'content-length': [u'Content
-Length', u'1132'], u'x-http-hmac': [u'x-http-hmac', u'oa9gbzO+MlRWiUoN07pxFp9A4m5GfKrwK+12WDUwh+c='], u'accept-encoding': [u'Accept-En
coding', u'gzip, deflate'], u'accept': [u'Accept', u'*/*'], u'user-agent': [u'User-Agent', u'python-requests/2.9.1'], u'connection': [u
'Connection', u'keep-alive'], u'content-type': [u'content-type', u'application/json']}, u'TYPE': u'CaseInsensitiveDict'}, u'hooks': {u'
response': []}, u'_cookies': {u'_now': 1495397816, u'_policy': {u'strict_rfc2965_unverifiable': True, u'strict_ns_domain': 0, u'_allowe
d_domains': None, u'rfc2109_as_netscape': None, u'rfc2965': False, u'strict_domain': False, u'_now': 1495397816, u'strict_ns_set_path':
 False, u'strict_ns_unverifiable': False, u'strict_ns_set_initial_dollar': False, u'hide_cookie2': False, u'_blocked_domains': [], u'TY
PE': u'instance', u'netscape': True}, u'_cookies_lock': {u'_Verbose__verbose': False, u'_RLock__owner': None, u'TYPE': u'_RLock', u'_RL
ock__block': u'<thread.lock object at 0x108ffa930>', u'_RLock__count': 0}, u'_cookies': {}, u'TYPE': u'RequestsCookieJar'}, u'method':
u'POST'}, u'TYPE': u'ConnectionError', u'response': None}, u'traceback': u'Traceback (most recent call last):\n  File "/Users/ben/.vim/
bundle/YouCompleteMe/third_party/ycmd/ycmd/../ycmd/handlers.py", line 103, in GetCompletions\n    .ComputeCandidates( request_data ) )\
n  File "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/../ycmd/completers/completer.py", line 217, in ComputeCandidates\n
   candidates = self._GetCandidatesFromSubclass( request_data )\n  File "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/../
ycmd/completers/completer.py", line 233, in _GetCandidatesFromSubclass\n    raw_completions = self.ComputeCandidatesInner( request_data
 )\n  File "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/../ycmd/completers/swift/swift_completer.py", line 263, in Compu
teCandidatesInner\n    for completion in self._FetchCompletions( request_data ) ]\n  File "/Users/ben/.vim/bundle/YouCompleteMe/third_p
arty/ycmd/ycmd/../ycmd/completers/swift/swift_completer.py", line 293, in _FetchCompletions\n    request_data )\n  File "/Users/ben/.vi
m/bundle/YouCompleteMe/third_party/ycmd/ycmd/../ycmd/completers/swift/swift_completer.py", line 196, in _GetResponse\n    headers = ext
ra_headers )\n  File "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/requests/requests/api.py", line 53, in request\
n    return session.request(method=method, url=url, **kwargs)\n  File "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_part
y/requests/requests/sessions.py", line 468, in request\n    resp = self.send(prep, **send_kwargs)\n  File "/Users/ben/.vim/bundle/YouCo
mpleteMe/third_party/ycmd/third_party/requests/requests/sessions.py", line 576, in send\n    r = adapter.send(request, **kwargs)\n  Fil
e "/Users/ben/.vim/bundle/YouCompleteMe/third_party/ycmd/third_party/requests/requests/adapters.py", line 426, in send\n    raise Conne
ctionError(err, request=request)\nConnectionError: (\'Connection aborted.\', BadStatusLine("\'\'",))\n', u'message': u'(\'Connection ab
orted.\', BadStatusLine("\'\'",))'}], u'completions': [], u'completion_start_column': 9}
Error detected while processing function <SNR>126_PollCompletion[15]..<SNR>126_Pyeval:
line    4:
E859: Failed to convert returned python object to vim value
Error detected while processing function <SNR>126_PollCompletion:
line   16:
E121: Undefined variable: completion_start_column
E15: Invalid expression: {   'start_column': response.completion_start_column,   'candidates': response.completions }
```

So i think the issue is that Vim can't deserialise the 'errors' thing to its internal format?

Tis change in `completion_request.py` fixes it:

```python

      for e in errors:
        with HandleServerException( truncate = True ):
          raise MakeServerException( e )

      if 'completions' in response and 'completion_start_column' in response:
        return {
          'completions': response[ 'completions' ],
          'completion_start_column': response[ 'completion_start_column' ]
        }
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 5 of 9 files at r1, 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Alright, thanks for sharing the code. I agree that users would most likely like to customise it, so leaving it out would be the best.

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 240 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkahLtsFej08Ze8tIc1:-KkhQwaTyAZO-vsmc7BX:b-nxqo04) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L240)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

I added `<C-h>` to the list of keys. I also added the `<silent>` attribute so that if the user has already mapped the key to something else, we don't override it.
</blockquote></details>

Sounds good to me. Just one note, it's `<unique>` not `<silent>`. The code is good, but your comment made me look up Vim documentation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KkhQ_7AiuEDlE1IQEG2:b-5snzwf)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc5cV3GhPHMyKI4wom:-KkhWH2Yg3ph8cjLVMn9:b2hdgtu) (no related file):*
> So i think the issue is that Vim can't deserialise the 'errors' thing to its internal format?

Yes, that's the issue. Fixed.

---

*[autoload/youcompleteme.vim, line 240 at r1](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KkahLtsFej08Ze8tIc1:-KkhVCHRJZYDgepfho6v:b-9ninsl) ([raw file](https://github.com/valloric/youcompleteme/blob/e008a4388577c58c40c1fbd92041f5d19f16b4a8/autoload/youcompleteme.vim#L240)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Sounds good to me. Just one note, it's `<unique>` not `<silent>`. The code is good, but your comment made me look up Vim documentation.
</blockquote></details>

Sorry, I mixed up the two because of the `silent!` at the start of the command.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Reviewed 1 of 1 files at r5, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-Kkc50bnWiBhe43Da6Aw:-KkcEr9yHlBbFWafCWss:bcuch00) (no related file):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

It seems to be fixed by leaving semantic mode and resetting completions before the first `s:Complete()` call. Could you try again with the new changes?
</blockquote></details>

I have tried the example above and it seems to work. I didn't get the semantic completion after entering the second space. The tests are passing as well, so this is <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> as far as I can see.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

Valloric commented:
This looks pretty good to me! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-Kl9mQldZ3fYGBJi2ukA:bp6onfe)*
<!-- Sent from Reviewable.io -->

micbou commented:
I fixed a few bugs and did some refactoring. I am going to give this PR one more week of testing before changing its status to ready (unless a major issue is found).

---

Reviewed 1 of 1 files at r5, 1 of 1 files at r6, 2 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 684 at r7](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KlFye3rAgaruu15E4_6:-KlFye3rAgaruu15E4_7:b53z2xs) ([raw file](https://github.com/valloric/youcompleteme/blob/cb83e992457b91e2102353a4f399c4969daf9a66/autoload/youcompleteme.vim#L684)):*
> ```vim
>     call feedkeys( "\<C-e>", 'n' )
>   endif
> endfunction
> ```

Removed this function as it was only used in `s:OnInsertChar`.

---

*[autoload/youcompleteme.vim, line 526 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KlFy09s4WZaQXMPq-UP:-KlFy09s4WZaQXMPq-UQ:btrbwej) ([raw file](https://github.com/valloric/youcompleteme/blob/60fb4351994a03181d19748c18b86abcf1781ff2/autoload/youcompleteme.vim#L526)):*
> ```vim
> 
> 
> function! s:OnInsertChar()
> ```

Renamed to be consistent with `OnDeleteChar`.

---

*[autoload/youcompleteme.vim, line 534 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KlFv7OKhGVqS0sH-wjT:-KlFv7OKhGVqS0sH-wjU:bx8c3z1) ([raw file](https://github.com/valloric/youcompleteme/blob/60fb4351994a03181d19748c18b86abcf1781ff2/autoload/youcompleteme.vim#L534)):*
> ```vim
> 
> 
> function! s:OnDeleteChar( key )
> ```

Renamed the parameter from `char` to `key` since it's actually a key (`<BS>` or `<C-h>`), not a character.

---

*[autoload/youcompleteme.vim, line 575 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KlFvk1eJ7qr8BXCrxwX:-KlFvk1eJ7qr8BXCrxwY:balt3ks) ([raw file](https://github.com/valloric/youcompleteme/blob/60fb4351994a03181d19748c18b86abcf1781ff2/autoload/youcompleteme.vim#L575)):*
> ```vim
>   endif
> 
>   if &completefunc == "youcompleteme#CompleteFunc" &&
> ```

I did remove this condition but it is actually needed to avoid the `E764: Option 'completefunc' is not set` error.

---

*[autoload/youcompleteme.vim, line 580 at r8](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-KlFy5kufigziGhB1Tsi:-KlFwhUrOnadKBoVs88Z:bqteyif) ([raw file](https://github.com/valloric/youcompleteme/blob/60fb4351994a03181d19748c18b86abcf1781ff2/autoload/youcompleteme.vim#L580)):*
> ```vim
>         \ !s:OnBlankLine()
>     " Immediately call previous completion to avoid flickers.
>     call s:Complete()
> ```

Don't display last completions if we are inside a comment/string or on a blank line. Fix a bug where completions were displayed after a C++-style comment `//`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KlFuACHRD3qFDn33Bn2:bdetbax)*
<!-- Sent from Reviewable.io -->

micbou commented:
With the new system, users will complain about the impossibility to close the completion menu with the `<C-y>` key. It's already an issue with the current system (see https://github.com/Valloric/YouCompleteMe/issues/1282) but it's way worse with the new one. We could remap the `<C-y>` key to really close the completion menu. We could even allow the user to change the key through an option. Should we implement a way to close the menu in this PR?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KlpOnsWH9EpHSjBH7di:bgtkig7)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
I believe that we should try to fix that issue here. It may be an edge case, but with our policy stating "every commit is a release", I think we should not break basic Vim functionality.

---

Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KlG7CGAQkPB3p37gSvL:bub80hi)*
<!-- Sent from Reviewable.io -->

micbou commented:
Added a new option `g:ycm_key_list_stop_completion` that contains a list of keys (only `<C-y>` by default) for closing the completion menu.

---

Reviewed 8 of 8 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-Kn76tPxZe7TOebVpz3Z:b2vz3ls)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 2 of 8 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
This seems to work fine, but should we do something about `<C-e>`? It's broken as well and it's purpose is not only to close the pop-up menu, but also discard the completion text and give the user back the orginal text the user has typed.

---

Reviewed 7 of 8 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KnF1yf5NRcJ9fYoqb8X:b6aua4w)*
<!-- Sent from Reviewable.io -->

micbou commented:
`<C-e>` still reverts to the original text. If the user want to close the menu as well, he can do that by typing `<C-e><C-y>`. I think this is good enough. In fact, I would even argue it's better than the default behavior because it's more flexible. You can either revert to the original text or close the completion menu. You can't do both separately by default. 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KnF713IBAIUxim-b19w:b-38rk3n)*
<!-- Sent from Reviewable.io -->

Valloric commented:
> <C-e> still reverts to the original text. If the user want to close the menu as well, he can do that by typing <C-e><C-y>. I think this is good enough.

I agree that this is good enough. Let's proceed then.

bstaletic commented:
> `<C-e>` still reverts to the original text. If the user want to close the menu as well, he can do that by typing `<C-e><C-y>`. I think this is good enough.

I'm fine with this solution as well. Is there anything elsse we need to take care of before merging this?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KnUZhdXo5kprD0fz6nz:ba95obn)*
<!-- Sent from Reviewable.io -->

micbou commented:
No, I think this is ready to be merged.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KnUzujn-6PwJRJlEivl:b2bcuvu)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
@zzbot r+

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2657#-:-KnV2mE8Uh5p-MGTnC2s:bi59kwd)*
<!-- Sent from Reviewable.io -->

sesselastronaut commented:
> autoload/youcompleteme.vim, line 575 at r8 (raw file):
> 
>       endif
> 
>       if &completefunc == "youcompleteme#CompleteFunc" &&
> 
> I did remove this condition but it is actually needed to avoid the E764: Option 'completefunc' is not set error.

Since I've updated the plug-in I'm affected by this, <C-Space> now throws the E764 error above. So far couldn't find any solution. I've  `g:ycm_auto_trigger` set to `0` as pointed out here [(https://github.com/Valloric/YouCompleteMe/issues/1634)] but this doesn't change the issue. Many thanks for any indication to find
[vimrc.txt](https://github.com/Valloric/YouCompleteMe/files/1155675/vimrc.txt)

 the fly in the ointment.

micbou commented:
@sesselastronaut That's because of these lines:
```viml
if has("gui_running")
  inoremap <C-Space> <C-x><C-o>
else
  inoremap <Nul> <C-x><C-o>
endif
```
in your vimrc. You don't need them anymore. See issue https://github.com/Valloric/YouCompleteMe/issues/2699.

sesselastronaut commented:
@micbou thanks for your response. I've removed those lines but still get the same behaviour featuring the E764: Option 'completefunc' is not set error. I've also tried the 
```
let g:ycm_key_invoke_completion = '<leader><leader>'
let g:ycm_key_list_stop_completion = ['<C-y>', '<CR>']
```
options from the linked threat but they do not change it either.

micbou commented:
What's the output of `:imap <C-Space>`?

sesselastronaut commented:
`i  <C-Space>   * <C-R>=<SNR>50_InvokeSemanticCompletion()<CR>`

micbou commented:
Thanks. I see the issue. We don't check if `completefunc` is set to `youcompleteme#CompleteFunc` when forcing semantic completion. I'll work on a fix. Still, I am wondering why it's not set in your case. What is the filetype of the current buffer:
```viml
:set filetype?
```
?

sesselastronaut commented:
it is a latex document and `:set filetype?` returns `filetype=tex`. 
Many thanks in advance for your effort to fix this!

sesselastronaut commented:
@micbou thanks for that, I've updated the plugin and the error is gone but unfortunately, the autocompletion with <C-Space> still doesn't work for me. 

I've tried setting`let g:ycm_key_invoke_completion = '<C-Space>'` which is afaik the default anyway. The output of `:imap <C-Space>` is `i  <C-Space>   * <C-R>=<SNR>95_InvokeSemanticCompletion()<CR>`. 

Don't know if this info is helpful to track this down but strangely enough with `let g:ycm_key_invoke_completion = '<C-Space>'` set/uncommented I've observed the following behaviour: when I open a latex file and type `\cite{` or `\cite{}`followed by `<C-Space>` nothing really happens. However leaving insert mode immediately results in `The ycmd server SHUT DOWN (restart w... the instructions in the documentation.` message.

micbou commented:
Could you paste the contents of the ycmd stderr logfile? You can open it directly in Vim with [the `:YcmToggleLogs` command](https://github.com/Valloric/YouCompleteMe#the-ycmtogglelogs-command). Use `<TAB>` to cycle through the logfiles.

sesselastronaut commented:
```
2017-08-02 16:19:20,636 - ERROR - ycm_core library not detected; you need to compile it by running the build.py script. See the documentation for more details.
Traceback (most recent call last):
  File "/home/olsen/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/server_utils.py", line 96, in CompatibleWithCurrentCore
    ycm_core = ImportCore()
  File "/home/olsen/.vim/bundle/YouCompleteMe/third_party/ycmd/ycmd/server_utils.py", line 88, in ImportCore
    import ycm_core as ycm_core
ImportError: No module named ycm_core
```

micbou commented:
Run the `install.py` script in `/home/olsen/.vim/bundle/YouCompleteMe` folder and try again.

sesselastronaut commented:
Ok, `cmake` was missing on my system. After installing it and running the `install.py` the completion with `<C-Space>` seems to be working now :) many thanks for tracking this down!
